azurelinuxagent/ga/guestagent.py

```diff
@@ -0,0 +1,316 @@
+import json
+import os
+import shutil
+import time
+
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.utils import textutil
+
+from azurelinuxagent.common import logger, conf
+from azurelinuxagent.common.exception import UpdateError
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.version import AGENT_DIR_PATTERN, AGENT_NAME, CURRENT_VERSION
+from azurelinuxagent.ga.exthandlers import HandlerManifest
+
+AGENT_ERROR_FILE = "error.json"  # File name for agent error record
+AGENT_MANIFEST_FILE = "HandlerManifest.json"
+MAX_FAILURE = 3  # Max failure allowed for agent before declare bad agent
+
+
+class GAUpdateReportState(object):
+    """
+    This class is primarily used to maintain the in-memory persistent state for the agent updates.
+    This state will be persisted throughout the current service run and might be modified by external classes.
+    """
+    report_error_msg = ""
+    report_expected_version = FlexibleVersion("0.0.0.0")
+
+
+class GuestAgent(object):
```

Comment:
We should split this class in code to download the agent and code to get the information of agents that are already downloaded. We can follow offline.